### PR TITLE
Shell Api to get default Application for a file type

### PR DESCRIPTION
**Pull request by niteskum**
API is added  to get Default Application for a file type.
Api Name: getSystemDefaultApp
This takes input as comma separated file extensions string and return comma separated OS Default Application for each extensions if any.

@swmitra @shubhsnov @sobisht @jha-g  @narayani28  please review

### CODE REVIEW

File: appshell/appshell_extensions_win.cpp [hunk 1]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+            appPath = appPath.substr(0, appPath.find(L"_"));
+            return true;
+        }
+        return false;
+    }
+    return true;
+
+}
+
+BOOL GetShellDefaultOpenWithProgPath(ExtensionString fileExt, ExtensionString &appPath)
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: const is missing.

File: appshell/appshell_extensions_win.cpp [hunk 2]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+        }
+
+        ::RegCloseKey(extKey);
+    }
+
+    return result;
+}
+
+
+BOOL GetLegacyWin32SystemEditor(ExtensionString fileExt, ExtensionString &appPath)
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: I think ExtensionString can be `const &`

File: appshell/appshell_extensions_win.cpp [hunk 3]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+        {
+            return ResolveAppPathFromProgID(subKeyStr, appPath);
+            }
+        }
+        RegCloseKey(extKey);
+
+    return FALSE;
+}
+
+int32 getSystemDefaultApp(ExtensionString fileTypes, ExtensionString& fileTypesWithdefaultApp)
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: I think ExtensionString can be `const &`

File: appshell/appshell_extensions_win.cpp [hunk 4]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+    ExtensionString separator = L"##";
+
+    wchar_t* token = std::wcstok((wchar_t*)fileTypes.c_str(), delim, &nextPtr);
+ 
+    while (token) {
+        extArray.push_back(token);
+        token = wcstok(NULL, delim, &nextPtr);
+    }
+
+    for (std::vector<ExtensionString>::iterator it = extArray.begin(); it != extArray.end(); ++it) {
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: Use a const iterator or 
for (auto const & item : extArray)
{
}
or std::for_each.

File: appshell/appshell_extensions_mac.mm [hunk 5]
```diff
@@ -391,6 +391,59 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+    char delim[] = ",";
+    char separator[] = "##";
+    std::vector<ExtensionString> extArray;
+    char* token = std::strtok((char*)fileTypes.c_str(), delim);
+    while (token) {
+        extArray.push_back(token);
+        token = std::strtok(NULL, delim);
+    }
+
+    for (std::vector<ExtensionString>::iterator it = extArray.begin(); it != extArray.end(); ++it)
```
Comment (jha-g) [inline appshell/appshell_extensions_mac.mm]: Use a const iterator or
for (auto const & item : extArray)
{
}
or std::for_each.

File: appshell/appshell_extensions_mac.mm [hunk 6]
```diff
@@ -391,6 +391,59 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+                {
+                    appPath = [(NSString *)[bundle objectForInfoDictionaryKey: @"CFBundleExecutable"] cStringUsingEncoding:NSUTF8StringEncoding];
+                }
+
+                fileTypesWithdefaultApp = fileTypesWithdefaultApp + *it + separator + appPath + ",";
+
+                CFRelease(bundle_id);
+            }
+
+            CFRelease(UTI);
```
Comment (jha-g) [inline appshell/appshell_extensions_mac.mm]: Can we  use scope based auto-release-pool.

File: appshell/appshell_extensions_win.cpp [hunk 7]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+    {
+        TCHAR subKeyStr[255];
+        ULONG subKeySize = 255;
+        DWORD index = 0;
+        LONG err = RegEnumValue(extKey, index, subKeyStr, &subKeySize,
+            0, 0, 0, NULL);
+        if (err != ERROR_NO_MORE_ITEMS)
+        {
+            return ResolveAppPathFromProgID(subKeyStr, appPath);
+            }
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: This brace looks dangling.

File: appshell/appshell_extensions_win.cpp [hunk 8]
```diff
@@ -403,6 +403,166 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
     }
 }
 
+
+static BOOL ResolveAppPathCommon(ExtensionString lpszKey, ExtensionString lpszVal, ExtensionString& appPath)
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: Can we make arguments const or const &.

File: appshell/appshell_extensions_win.cpp [hunk 9]
```diff
@@ -403,6 +403,170 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+}
+
+BOOL GetShellDefaultOpenWithProgPath(const ExtensionString& fileExt, ExtensionString &appPath)
+{
+    HKEY	extKey = NULL;
+    ExtensionString key = fileExt;
+    BOOL	result = FALSE;
+    ULONG	regKey(REG_SZ);
+
+    if (fileExt[0] != '.')
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: fileExt can be empty, prefer fileExt.front().

File: appshell/appshell_extensions_win.cpp [hunk 10]
```diff
@@ -403,6 +403,170 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+    }
+
+    return result;
+
+}
+
+static BOOL ResolveAppPathFromProgID(const ExtensionString& lpszProgID, ExtensionString& appPath)
+{
+
+    ExtensionString key;
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: Why split one statement in two line.

File: appshell/appshell_extensions_win.cpp [hunk 11]
```diff
@@ -403,6 +403,170 @@ void CloseLiveBrowser(CefRefPtr<CefBrowser> browser, CefRefPtr<CefProcessMessage
...
+}
+
+
+BOOL GetLegacyWin32SystemEditor(const ExtensionString& fileExt, ExtensionString &appPath)
+{
+    HKEY	rootKey;
+    ULONG	regKey(REG_SZ);
+    ExtensionString key = fileExt;
+
+    if (fileExt[0] != '.')
```
Comment (jha-g) [inline appshell/appshell_extensions_win.cpp]: prefer fileExt.front()

**Comment by jha-g** [comment]
Address these minor issues, rest looks good.

File: appshell/appshell_extensions_win.cpp (hunk 9 above)
Comment (niteskum) [reply appshell/appshell_extensions_win.cpp]: if "fileext" is empty, it should return false. added code above.

File: appshell/appshell_extensions_win.cpp (hunk 11 above)
Comment (niteskum) [reply appshell/appshell_extensions_win.cpp]: if "fileext" is empty, it should return false. added code above.

**Comment by niteskum** [comment]
made the suggested changes

**Comment by jha-g** [comment]
LGTM

**Pull request merged**